crud/common/sharding/sharding_func.lua

```diff
@@ -93,6 +93,31 @@ function sharding_func_module.extract_function_def(tuple)
     return nil
 end
 
+function sharding_func_module.construct_as_callable_obj_cache(metadata_map, specified_space_name)
+    dev_checks('table', 'string')
+
+    local result_err
+
+    cache.sharding_func_map = {}
+    for space_name, metadata in pairs(metadata_map) do
+        if metadata.sharding_func_def ~= nil then
+            local sharding_func, err = as_callable_object(metadata.sharding_func_def,
+                                                          space_name)
+            if err ~= nil then
+                if specified_space_name == space_name then
+                    result_err = err
+                else
+                    log.warn(err)
+                end
+            end
+
+            cache.sharding_func_map[space_name] = sharding_func
```

Comment:
Let's consider the case, when the function is defined by its name (strictly speaking by a path in the dot notation) in `_ddl_sharding_func`. If we'll update the function itself (the callable object, e.g. `_G.my_sharding_func = function(key) <...> end`), the old callable will stuck in the cache.

We're solving the same problem in https://github.com/tarantool/ddl/issues/82 (consider 6th bullet).

That's lowers performance of the bucket id calculation in 1.5 times for strcrc32 alike function according to numbers in https://github.com/tarantool/ddl/pull/85. But it is the safe default behaviour.

I suggest to look at the cache strcture in the linked pull request: here we store either parsed dot notation (`{'vshard', 'router', 'mpcrc32'}` for example) or callable (when a function is set as a body).

Reply:
I would make exception for `vshard.router.bucket_id_strcrc32()`, `vshard.router.bucket_id()` and `vshard.router.bucket_id_mpcrc32()` by default (they are known as stable) and allow users to manage those exceptions.

Since most usages are about built-in vshard bucket id functions, your current solution looks good enough. Because of that and because it is simpler, I retire my comment (but vote for highlighting the limitation in readme).

Reply:
Filed related issue regarding tarantool/ddl: https://github.com/tarantool/ddl/issues/89.